cubi_tk/sodar/ingest.py

```diff
@@ -0,0 +1,366 @@
+"""``cubi-tk sodar ingest``: add arbitrary files to SODAR"""
+
+import argparse
+import os
+import sys
+from pathlib import Path
+import getpass
+import typing
+import attr
+
+from irods.session import iRODSSession
+from logzero import logger
+from sodar_cli import api
+from ..common import load_toml_config, is_uuid, compute_md5_checksum, sizeof_fmt
+from ..snappy.itransfer_common import TransferJob
+from tqdm import tqdm
+
+
+@attr.s(frozen=True, auto_attribs=True)
+class Config:
+    """Configuration for the ingest command."""
+
+    config: str = attr.field(default=None)
+    sodar_server_url: str = attr.field(default=None)
+    sodar_api_token: str = attr.field(default=None, repr=lambda value: "***")  # type: ignore
+
+
+class SodarIngest:
```

Comment:
I think most of the functionality here should - ideally - be part of an abstract base class, similar to SnappyItransferCommandBase . Then both a general filetype agnostic Ingest could be derived from it and it could also be used for things like SodarIngestFastq (where the reliance on some snappy modules that load/assume biomedsheet like samplesheets is porbably not ideal anyway). 

I'm not sure how important that change is right now though, even if it's the best long term approach

Reply:
I've extracted much of the iRODS logic into cubi_tk/irods_utils.py. Please have a look.